File: F/ioutils.f90
subroutine createfilename(filename, basename, nx, ny, rank)

  implicit none

  character*(*) :: filename, basename

  integer :: nx, ny, rank

  if (rank .lt. 0) then

    write(filename, fmt='(i10.10,''x'',i10.10,''.dat'')') nx, ny

  else

    write(filename, fmt='(i10.10,''x'',i10.10,''_'',i4.4,''.dat'')')  nx, ny, rank

  end if

  filename = basename//filename

end

subroutine iosize(filename, nx, ny)

  implicit none

  character*(*) :: filename
  integer       :: nx, ny

  integer :: i

  do i = 1, len(filename)

    if (iachar(filename(i:i)) .ge. iachar('0') .and. &
        iachar(filename(i:i)) .le. iachar('9')         ) exit

  end do

  if (i > len(filename)) then

    write(*,*) 'iosize: error parsing filename ', filename

    nx = -1
    ny = -1

  else

    read(filename(i:len(filename)), fmt='(i4.4,''x'',i4.4,''.dat'')') nx, ny

  end if

end



subroutine ioread(filename, data, nreal)

  implicit none

  integer, parameter :: iounit   = 10
  integer, parameter :: realsize = 4

  character*(*) :: filename
  integer       :: nreal
  real          :: data(nreal)

  integer :: i

  write(*,*) 'ioread: reading ', filename

  open(unit=iounit, file=filename, form='unformatted', &
       access='direct', recl=realsize)

  do i = 1, nreal
    read(unit=iounit, rec=i) data(i)
  end do

  close(unit=iounit)

  write(*,*) '.. done'

end

subroutine iochunkread(filename, data, nreal, offset)

  implicit none

  integer, parameter :: iounit   = 10
  integer, parameter :: realsize = 4

  character*(*) :: filename
  integer       :: nreal
  real          :: data(nreal)
  integer(kind=8)::offset

  integer(kind=8) :: i

!AJ remove large amounts of output for large files
!AJ  write(*,*) 'iochunkread: reading ', filename

  open(unit=iounit, file=filename, form='unformatted', &
       access='direct', recl=realsize)

  do i = 1,nreal     
    read(unit=iounit, rec=i+offset) data(i)
  end do

  close(unit=iounit)

!AJ remove large amount of output for large files
!AJ  write(*,*) '.. done'

end


subroutine iowrite(filename, data, nreal)

  implicit none

  integer, parameter :: iounit   = 10
  integer, parameter :: realsize = 4

  character*(*) :: filename
  integer       :: nreal
  real          :: data(nreal)

  integer :: i

  write(*,*) 'iowrite: writing ', filename

  open(unit=iounit, file=filename, form='unformatted', &
       access='direct', recl=realsize)

  do i = 1, nreal
    write(unit=iounit, rec=i) data(i)
  end do

  close(unit=iounit)

  write(*,*) '.. done'

end

subroutine initarray(data, nx, ny)

  implicit none

  real, parameter :: initdataval = 0.5

  integer :: nx, ny

  real data(nx*ny)

  integer :: i

  do i = 1, nx*ny
    data(i) = initdataval
  end do

end

subroutine initpgrid(pcoords, nxproc, nyproc)

  use mpi

  implicit none

  integer, parameter :: ndim = 2

  integer :: nxproc, nyproc
  integer, dimension(ndim, nxproc*nyproc) :: pcoords

  integer, dimension(ndim) :: dims, periods

  integer :: i, ierr
  integer :: comm = MPI_COMM_WORLD
  integer :: gridcomm

  logical :: reorder

  periods = (/ 0, 0 /)
  reorder = .false.

  dims(1) = nxproc
  dims(2) = nyproc

  call MPI_CART_CREATE(comm, ndim, dims, periods, reorder, gridcomm, ierr)

  do i = 1, nxproc*nyproc

    call MPI_CART_COORDS(gridcomm, i-1, ndim, pcoords(1, i), ierr)

  end do

  call MPI_COMM_FREE(gridcomm, ierr)

end subroutine initpgrid

subroutine checkandgetargumentssub(filename, maxfilename, nx, ny, xprocs, yprocs, nxp, nyp, subnum, ioworker, lowerrank, upperrank, barrier, size, rank)

  use mpi

  implicit none

  integer, intent(in) :: size, rank, maxfilename
  integer, intent(out) :: nx, ny, xprocs, yprocs, nxp, nyp, barrier, subnum, ioworker, lowerrank, upperrank
  character*(maxfilename), intent(out) :: filename
  integer :: i, numargs, chunk, ierr
  character(len=32) :: arg


  numargs = command_argument_count()
  if(numargs .ne. 7) then
     if(rank .eq. 0) then
        write(*,*) "usage: inputfilename nx ny xprocs yprocs barrier numio"
        write(*,*) "This application expects you to provide the input file name, the size of the input data set (nx*ny),"
        write(*,*) "the number of processes you want in each dimension (xprocs and yprocs), and"
        write(*,*) "whether to use a barrier before timing (barrier, 0 is no barrier, 1 uses a barrier)"
     end if
     call MPI_FINALIZE(ierr)
     stop
  else
     call getargs(filename, maxfilename, nx, ny, xprocs, yprocs, nxp, nyp, barrier, size, rank)
  end if
  call getarg(7,arg)
  arg = trim(arg)
  read(arg, '(I10)') subnum
  if((size/subnum)*subnum .ne. size) then
     if(rank .eq. 0) then
        write(*,*) 'The number of I/O workers specified',subnum,'does not exactly divide the number of processes used',size
     end if
     call MPI_FINALIZE(ierr)
     stop
  end if
  chunk = size/subnum
  if(mod(rank,chunk) .eq. 0) then
     ioworker = 1
     lowerrank = rank
     upperrank = rank + chunk - 1
  else
     ioworker = 0
     lowerrank = rank-(mod(rank,chunk))
     upperrank = rank-(mod(rank,chunk))
  end if

  if(rank .eq. 0) then
     write(*,*) 'Number of I/O workers:',subnum
  end if

end subroutine checkandgetargumentssub

subroutine checkandgetarguments(filename, maxfilename, nx, ny, xprocs, yprocs, nxp, nyp, barrier, size, rank)

  use mpi

  implicit none

  integer, intent(in) :: size, rank, maxfilename
  integer, intent(out) :: nx, ny, xprocs, yprocs, nxp, nyp, barrier
  character*(maxfilename), intent(out) :: filename
  integer :: i, numargs, ierr
  character(len=32) :: arg

  numargs = command_argument_count()
  if(numargs .ne. 6) then
     if(rank .eq. 0) then
        write(*,*) "usage: inputfilename nx ny xprocs yprocs barrier"
        write(*,*) "This application expects you to provide the input file name, the size of the input data set (nx*ny),"  
        write(*,*) "the number of processes you want in each dimension (xprocs and yprocs), "
        write(*,*) "and whether to use a barrier before timing (barrier, 0 is no barrier, 1 uses a barrier)"
     end if
     call MPI_FINALIZE(ierr)
     stop
  else
     call getargs(filename, maxfilename, nx, ny, xprocs, yprocs, nxp, nyp, barrier, size, rank)
  end if

end subroutine checkandgetarguments

subroutine getargs(filename, maxfilename, nx, ny, xprocs, yprocs, nxp, nyp, barrier, size, rank)

  use mpi

  implicit none

  integer, intent(in) :: size, rank, maxfilename
  integer, intent(out) :: nx, ny, xprocs, yprocs, nxp, nyp, barrier
  character*(maxfilename), intent(out) :: filename

  integer :: i, numargs, ierr
  character(len=32) :: arg
 
  call getarg(1,filename)
  call getarg(2,arg)
  arg = trim(arg)
  read(arg, '(I10)') nx
  call getarg(3,arg)
  arg = trim(arg)
  read(arg, '(I10)') ny
  call getarg(4,arg)
  arg = trim(arg)
  read(arg, '(I10)') xprocs
  call getarg(5,arg)
  arg = trim(arg)
  read(arg, '(I10)') yprocs
  call getarg(6,arg)
  arg = trim(arg)
  read(arg, '(I2)') barrier
  if(xprocs*yprocs .ne. size) then
     if(rank .eq. 0) then
        write(*,*) 'The specified xprocs and yprocs assignment does not match the total number of processes being used.'
        write(*,*) 'xprocs is ',xprocs,' yprocs is ',yprocs,' but total processes used is ',size,' which does not match xprocs * yprocs'
     end if
     call MPI_FINALIZE(ierr)
     stop
  end if
  nxp = nx/xprocs
  if(xprocs*nxp .ne. nx) then
     if(rank .eq. 0) then
        write(*,*) 'nx does not exactly divide by xprocs.  Stopping the code!'
     end if
     call MPI_FINALIZE(ierr)
     stop
  end if
  nyp = ny/yprocs
  if(yprocs*nyp .ne. ny) then
     if(rank .eq. 0) then
        write(*,*) 'ny does not exactly divide by yprocs. Stopping the code!'
     end if
     call MPI_FINALIZE(ierr)
     stop
  end if
  if(barrier .ne. 0 .and. barrier .ne. 1) then
     if(rank .eq. 0) then
        write(*,*) 'barrier should be 1 (use a barrier prior to starting the run) or 0 (do not use a barrier)'
     end if
     call MPI_FINALIZE(ierr)
     stop
  end if
  if(rank .eq. 0) then
     write(*,*) 'Running on',size,'processes'
     write(*,*) 'nx:',nx,'ny:',ny,'nxp:',nxp,'nyp:',nyp,'xprocs:',xprocs,'yprocs:',yprocs
     write(*,*) 'barrier:',barrier
  end if
  
end subroutine getargs


subroutine dotimings(totaltime, rank, size)

  use mpi

  implicit none

  double precision, intent(in) :: totaltime
  integer, intent(in) :: rank, size
  double precision :: avtotaltime, mintotaltime, maxtotaltime
  integer :: ierr

  call MPI_Reduce(totaltime, avtotaltime, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD, ierr)

  avtotaltime = avtotaltime/size

  call MPI_Reduce(totaltime, maxtotaltime, 1, MPI_DOUBLE, MPI_MAX, 0, MPI_COMM_WORLD, ierr)
  call MPI_Reduce(totaltime, mintotaltime, 1, MPI_DOUBLE, MPI_MIN, 0, MPI_COMM_WORLD, ierr)

  if(rank .eq. 0) then
   write(*,*) "Times (seconds): Average: ",avtotaltime," Maximum: ",maxtotaltime," Minimum: ",mintotaltime
  end if

end subroutine dotimings

File: C/ioutils.c
#include <stdio.h>
#include <stdlib.h>
#include "ioutils.h"
#include "mpi.h"


void checkandgetargumentssub(int argc, char **argv, int *nx, int *ny, int *xprocs, int *yprocs, int *nxp, int *nyp, int *subnum, int *ioworker, int *lowerrank, int *upperrank, int *barrier, int size, int rank){
  
  int chunk;

  if ( argc != 8 ){
    /* We print argv[0] assuming it is the program name */
    if(rank == 0){
      printf("usage: %s filename nx ny xprocs yprocs barrier numio", argv[0] );
      printf("This application expects you to provide the input file now, the size of the input data set (nx*ny), the number of processes you want in each dimension (xproc and yproc), and whether to use a barrier before timing (barrier, 0 is no barrier, 1 uses a barrier), the number of I/O workers you want (numio)\n");
    }
    MPI_Finalize();
    exit(-1);
    
  }else{
    getargs(argc,argv,nx,ny,xprocs,yprocs,nxp,nyp,barrier,size,rank);
    *subnum = atoi(argv[7]);
    if((size/(*subnum))*(*subnum) != size){
      if(rank == 0){
	printf("The number of I/O workers specified (%d) does not exactly divide the number of processes being used (%d)", *subnum, size);
      }
      MPI_Finalize();
      exit(-1);      
    }
    // Work out how many processes each I/O worker is reponsible for
    chunk = size/(*subnum);
    // Work out if this process is an I/O worker
    if(rank%chunk == 0){
      *ioworker = 1;
      // If we are an I/O worker then work out the range of ranks we are responsible for (including itself)
      *lowerrank = rank;
      *upperrank = rank + chunk - 1;
    
    }else{
      *ioworker = 0;
      // If we are not an I/O worker then record which process is our responsible I/O worker
      *lowerrank = rank-(rank%chunk);
      *upperrank = rank-(rank%chunk);
    }

  }
  if(rank == 0){
     printf("number of i/o workers: %d\n",*subnum);
  }
}

void checkandgetarguments(int argc, char **argv, int *nx, int *ny, int *xprocs, int *yprocs, int *nxp, int *nyp, int *barrier, int size, int rank){

  if ( argc != 7 ){
    /* We print argv[0] assuming it is the program name */
    if(rank == 0){
      printf("usage: %s filename nx ny xprocs yprocs barrier", argv[0] );
      printf("This application expects you to provide the input file now, the size of the input data set (nx*ny), the number of processes you want in each dimension (xproc and yproc), and whether to use a barrier before timing (barrier, 0 is no barrier, 1 uses a barrier)\n");
    }
    MPI_Finalize();
    exit(-1);
    
  }else{
    getargs(argc,argv,nx,ny,xprocs,yprocs,nxp,nyp,barrier,size,rank);
  }

}
void getargs(int argc, char **argv, int *nx, int *ny, int *xprocs, int *yprocs, int *nxp, int *nyp, int *barrier, int size, int rank){

  *nx = atoi(argv[2]);
  *ny = atoi(argv[3]);
  *xprocs = atoi(argv[4]);
  *yprocs = atoi(argv[5]);
  *barrier = atoi(argv[6]);
  
  if((*xprocs)*(*yprocs) != size){
    if(rank == 0){
      printf("Number of processes being used (%d) does not match number requested in the two dimension (%d x %d)\n",size,*xprocs,*yprocs);
    }
    MPI_Finalize();
    exit(-1);
  }
  
  *nxp = (*nx)/(*xprocs);
  *nyp = (*ny)/(*yprocs);
  
  if((*nxp)*(*xprocs) != *nx){
    if(rank == 0){
      printf("Number of processes in x dimension (xprocs) does not exactly divide the x dimension of the input (nx).  Quitting\n");
    }
    MPI_Finalize();
    exit(-1);
  }
  
  if((*nyp)*(*yprocs) != *ny){
    if(rank == 0){
      printf("Number of processes in y dimension (yprocs) does not exactly divide the y dimension of the input (ny).  Quitting\n");
    }
    MPI_Finalize();
    exit(-1);
  }
  
  if(*barrier != 0 && *barrier != 1){
    if(rank == 0){
      printf("Barrier should be 0 or 1\n");
    }
    MPI_Finalize();
    exit(-1);
  }

  if(rank == 0){
    printf("Running on %d processes\n",size);
    printf("nx: %d ny: %d nxp: %d nyp: %d xprocs: %d yprocs: %d\n",*nx,*nx,*nxp,*nyp,*xprocs,*yprocs);
    printf("barrier: %d\n",*barrier);
  }

}

void dotimings(double totaltime, int rank, int size){
  
  double avtotaltime, mintotaltime, maxtotaltime;

  MPI_Reduce(&totaltime, &avtotaltime, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);

  avtotaltime = avtotaltime/size;

  MPI_Reduce(&totaltime, &maxtotaltime, 1, MPI_DOUBLE, MPI_MAX, 0, MPI_COMM_WORLD);
  MPI_Reduce(&totaltime, &mintotaltime, 1, MPI_DOUBLE, MPI_MIN, 0, MPI_COMM_WORLD);

  if(rank == 0){
    printf("Times (seconds): Average: %lf Maximum: %lf Minimum: %lf\n",avtotaltime,maxtotaltime,mintotaltime);
  }


}

void createfilename(char *filename, char *basename, int nx, int ny, int rank)
{
  if (rank < 0)
    {
      sprintf(filename, "%s%04dx%04d.dat", basename, nx, ny);
    }
  else
    {
      sprintf(filename, "%s%04dx%04d_%02d.dat", basename, nx, ny, rank);
    }
}


void iosize(char *filename, int *nx, int *ny)
{ 
  char *tmpname;

  tmpname = filename;

  while (*tmpname < '0' || *tmpname > '9')
  {
    tmpname++;
  }

  if (sscanf(tmpname, "%04dx%04d.dat", nx, ny) != 2)
  {
    printf("iosize: error parsing filename <%s>\n", filename);
    exit(1);
  }
}


void ioread(char *filename, void *ptr, int nfloat)
{
  int i;

  FILE *fp;
  
  float *data = (float *) ptr;

  printf("ioread: reading <%s> ...\n", filename);

  if ( (fp = fopen(filename, "r")) == NULL)
  {
    printf("ioread: failed to open input file <%s>\n", filename);
    exit(1);
  }

  if (fread(data, sizeof(float), nfloat, fp) != nfloat) 
  {
    printf("ioread: error reading input file <%s>\n", filename);
    exit(1);
  }

  fclose(fp);

  printf("... done\n", filename);
}

void iochunkread(char *filename, void *ptr, int nfloat, long int offset)
{
  
  FILE *fp;
  
  float *data = (float *) ptr;

//  printf("iochunkread: reading <%s> ...\n", filename);

  if ( (fp = fopen(filename, "r")) == NULL)
  {
    printf("iochunkread: failed to open input file <%s>\n", filename);
    exit(1);
  }

  /* Initial datsa offset (moving to the correct starting point in the file */
  if (fseek(fp, offset, SEEK_SET) != 0){
    printf("iochunkread: failed to move the the correct place in the input file <%s>\n", filename);
    exit(1);
  }
  
  if (fread(data, sizeof(float), nfloat, fp) != nfloat) 
    {
      printf("iochunkread: error reading input file <%s>\n", filename);
      exit(1);
    }

  fclose(fp);

//  printf("... done\n", filename);
}

void iowrite(char *filename, void *ptr, int nfloat)
{
  int i;

  FILE *fp;
  
  float *data = (float *) ptr;

  printf("iowrite: writing <%s> ...\n", filename);

  if ( (fp = fopen(filename, "w")) == NULL)
  {
    printf("iowrite: failed to open output file <%s>\n", filename);
    exit(1);
  }

  if (fwrite(data, sizeof(float), nfloat, fp) != nfloat) 
  {
    printf("iowrite: error writing output file <%s>\n", filename);
    exit(1);
  }

  fclose(fp);
  printf("... done\n", filename);
}



void iochunkwrite(char *filename, void *ptr, int nfloat, long int offset)
{
  int i;

  FILE *fp;
  
  float *data = (float *) ptr;

  printf("iochunkwrite: writing <%s> ...\n", filename);

  if ( (fp = fopen(filename, "w")) == NULL)
  {
    printf("iochunkwrite: failed to open output file <%s>\n", filename);
    exit(1);
  }

  if (fseek(fp, offset, SEEK_SET) != 0){
    printf("iochunkread: failed to move the the correct place in the input file <%s>\n", filename);
    exit(1);
  }

  if (fwrite(data, sizeof(float), nfloat, fp) != nfloat) 
  {
    printf("iowchunkrite: error writing output file <%s>\n", filename);
    exit(1);
  }

  fclose(fp);
  printf("... done\n", filename);
}


#define INITDATAVAL 0.5

void initarray(void *ptr, int nx, int ny)
{
  int i, j;

  float *data = (float *) ptr;

  for (i=0; i < nx*ny; i++)
    {
      data[i] = INITDATAVAL;
    }
}

#define NDIM 2

void initpgrid(void *ptr, int nxproc, int nyproc)
{
  MPI_Comm gridcomm;
  MPI_Comm comm = MPI_COMM_WORLD;

  int dims[NDIM];
  int periods[NDIM] = {0, 0};

  int reorder = 0;

  int *pcoords = (int *) ptr;

  int i;

  dims[0] = nxproc;
  dims[1] = nyproc;

  MPI_Cart_create(comm, NDIM, dims, periods, reorder, &gridcomm);

  for (i=0; i < nxproc*nyproc; i++)
    {
      MPI_Cart_coords(gridcomm, i, NDIM, pcoords+NDIM*i);
    }

  MPI_Comm_free(&gridcomm);

}
